### ISSUE TITLE
[MRTCore] _purecall abort when retrieving QualifierValues of embedded file

### ISSUE THREAD

**Issue by MarkIngramUK** [microsoft/WindowsAppSDK#1238]
I'm conducting a test of embedded files when I ran into a crash. 

![image](https://user-images.githubusercontent.com/4758079/129388682-9f6c8d9b-558f-470f-8fe4-a3275173994a.png)

The line of code that is crashing:

```
RETURN_IF_FAILED(resolver->EvaluateDecision(&decision, &resultIndex, &qualifierSet));
```

The `__vfptr` of `resolver` doesn't look too good, but I don't know if that's just because I can't see everything, due to debug.

![image](https://user-images.githubusercontent.com/4758079/129388974-02976d78-441c-4732-be88-bf4dee9689cd.png)

My (partial) config is as follows:

```
  <index root="resources" startIndexAt="test">
    <default>
      <qualifier name="Language" value="en-GB"/>
      <qualifier name="Contrast" value="standard"/>
      <qualifier name="Scale" value="100"/>
      <qualifier name="HomeRegion" value="001"/>
      <qualifier name="TargetSize" value="256"/>
      <qualifier name="LayoutDirection" value="LTR"/>
      <qualifier name="Theme" value="dark"/>
      <qualifier name="AlternateForm" value=""/>
      <qualifier name="DXFeatureLevel" value="DX9"/>
      <qualifier name="Configuration" value=""/>
      <qualifier name="DeviceFamily" value="Universal"/>
      <qualifier name="Custom" value=""/>
    </default>
    <indexer-config type="folder" foldernameAsQualifier="true" filenameAsQualifier="true" qualifierDelimiter="."/>
    <indexer-config type="EmbedFiles"/>
  </index>
```

Here's the dump of the `.pri`:

```
<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<PriInfo>
	<ResourceMap name="redacted" version="1.0" primary="true">
		<Qualifiers>
			<Language>en-GB</Language>
		</Qualifiers>
		<ResourceMapSubtree name="Files">
			<NamedResource name="strings.txt" uri="ms-resource://redacted/Files/strings.txt">
				<Candidate qualifiers="Language-en-GB" isDefault="true" type="EmbeddedData">
					<Base64Value>[Embedded Binary Data]</Base64Value>
				</Candidate>
			</NamedResource>
		</ResourceMapSubtree>
	</ResourceMap>
</PriInfo>
```

And finally the lines of code that trigger it:

```
auto candidate = manager.MainResourceMap().GetValue(L"Files/strings.txt");
auto kind = candidate.Kind();
auto bytes = candidate.ValueAsBytes();
auto qualifierValues = candidate.QualifierValues(); // <--- here
```

**Comment by huichen123** [comment]
do you have a repro'ing project?

**Comment by MarkIngramUK** [comment]
I've created a small reproducible project. Extract the zip, open `MRTCrash.sln` and build and run (x64).

[MRTCrash.zip](https://github.com/microsoft/WindowsAppSDK/files/6990741/MRTCrash.zip)

**Comment by MarkIngramUK** [comment]
Great work, thanks @huichen123 :-)

**Comment by huichen123** [comment]
Thanks @MarkIngramUK for reporting the issue. You just exploited a deficiency in the tests, which missed testing ResourceCandidate.QualifierValues with null ResourceContext.

**Comment by MarkIngramUK** [comment]
@huichen123 , I've just discovered another related crash. I don't know if it's been fixed by #1251 or not, would it be possible for you to test the new scenario? I can create a new bug report if necessary.

Take the test app I attached above (https://github.com/microsoft/WindowsAppSDK/issues/1238#issuecomment-899313977) and change the body of the test to be:

```
    auto factory = winrt::get_activation_factory<ResourceManager, IResourceManagerFactory>();
    auto manager = factory.CreateInstance(L"strings.pri");
    
    ResourceCandidate candidate = nullptr;

    // If the context goes out of scope, it will crash when accessing qualifier values
    {
        ResourceContext context = manager.CreateResourceContext();
        candidate = manager.MainResourceMap().GetValue(L"Files/strings.txt", context);
    }

    auto qualifierValues = candidate.QualifierValues(); // Should crash here...
```

The `ResourceCandidate` doesn't contain a reference to the `ResourceContext`, so when the context goes out of scope, then later you try to access the qualifier values, you'll get a similar `_purecall` crash.
